core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1325,14 +1325,47 @@ private[spark] class BlockManager(
     blockInfoManager.releaseAllLocksForTask(taskAttemptId)
   }
 
+  /**
+   * Retrieve the given rdd block if it exists and is visible, otherwise call the provided
+   * `makeIterator` method to compute the block, persist it, and return its values.
+   *
+   * @return either a BlockResult if the block was successfully cached, or an iterator if the block
+   *         could not be cached.
+   */
+  def getOrElseUpdateRDDBlock[T](
+      taskId: Long,
+      blockId: RDDBlockId,
+      level: StorageLevel,
+      classTag: ClassTag[T],
+      makeIterator: () => Iterator[T]): Either[BlockResult, Iterator[T]] = {
+    val isCacheVisible = isRDDBlockVisible(blockId)
+    var computed: Boolean = false
+    val getIterator = () => {
+      computed = true
+      makeIterator()
+    }
+
+    val res = getOrElseUpdate(blockId, level, classTag, getIterator)
+    if (res.isLeft && !isCacheVisible) {
+      if (!computed) {
+        // Loaded from cache, re-compute to update accumulators.
+        makeIterator()
+      }
```

Comment:
If a block is not visible, and it was not computed in this task, we should not be using that block at all - even if cached here or elsewhere.
Particularly for unordered or indeterminate computation, this can impact the accumulators generated - and it wont match the result returned.

So we cant simply call `makeIterator` here - as @Ngone51 suggested, we should pass the flag to `getOrElseUpdate` and decide there based on visibility.

Reply:
> Particularly for unordered or indeterminate computation, this can impact the accumulators generated - and it wont match the result iterator returned.

There could be some other issues with such scenarios, like if we have blocks generated from different task attempts or cache re-compute, each cache may also have different data. Would this also be a problem?

Reply:
This is a good point - and couple of issues that come to mind:

* We do not disable speculative execution for indeterminate computation currently - and data generated from two task attempts can vary.
* We do not invalidate previously cached data, when there is a stage re-execution due to failures for an indeterminate computation,
This is something we should revisit @Ngone51, @jiangxb1987 - thoughts ?

Reply:
> We do not disable speculative execution for indeterminate computation currently - and data generated from two task attempts can vary (which could be cached, and so differ for same partition)

Right..in this case, the rdd block locations for different data can be attached to the same rdd block id. So the reader could get the different data for the same rdd block, which makes the rdd block data also indeterminate.

 > We do not invalidate previously cached data, when there is a stage re-execution due to failures for an indeterminate computation,

This seems to be a missing point in the indeterminate computation framework. @cloud-fan could you help confirm?

Reply:
> Right..in this case, the rdd block locations for different data can be attached to the same rdd block id. So the reader could get the different data for the same rdd block, which makes the rdd block data also indeterminate.

@Ngone51 do you think shall we solve this issue in this PR? This looks like a more general issue about the indeterminate computation.

Reply:
No. We should fix that issue separately.

Reply:
Btw, can you add a todo for this here @ivoson ? Thx

Reply:
Updated.